lib/code-caller-docker.js

```diff
@@ -0,0 +1,685 @@
+     * NOTE: in production, this will block code execution if the image isn't already
+     * present locally. Every effort to be made to ensure that the current image has
+     * already been pulled before PrairieLearn starts up.
+     */
+    async _ensureImage() {
+        const image = docker.getImage(getExecutorImageName());
+        try {
+            await image.inspect();
+        } catch (e) {
+            if (e.statusCode === 404) {
+                await docker.pull(getExecutorImageName(), {});
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Creates a container and attaches its stdin/stdout/stderr to streams
+     * we can write to and read from.
+     * 
+     * @param {string[]} binds - The bind mounts for the container
+     */
+    async _createAndAttachContainer(binds) {
+        this.container = await docker.createContainer({
+            name: `prairielearn.worker.${this.uuid}`,
+            Image: getExecutorImageName(),
+            OpenStdin: true,
+            // This will close stdin once we disconnect, which will let
+            // the worker know that they should die. Once the worker dies,
+            // `AutoRemove: true` below will ensure the worker container
+            // is removed.
+            StdinOnce: true,
+            AttachStdin: true,
+            AttachStdout: true,
+            AttachStderr: true,
+            HostConfig: {
+                Binds: binds,
+                AutoRemove: true,
+            },
+        });
+
+        const stream = await this.container.attach({
+            stream: true,
+            stdin: true,
+            stdout: true,
+            stderr: true,
+        });
+        this.stdinStream = new MemoryStream();
+        this.stdoutStream = new MemoryStream();
+        this.stderrStream = new MemoryStream();
+        this.stdinStream.pipe(stream);
+        this.container.modem.demuxStream(stream, this.stdoutStream, this.stderrStream);
+        this.stdoutStream.on('data', data => this._handleOutput(data));
+        // To see stderr from the container for debugging, uncomment the following line
+        // this.stderrStream.pipe(process.stderr);
+
+        await this.container.start();
+
+        // This will passively wait for the container to exit or be killed
+        this.container
+            .wait()
+            .then((status) => this._handleContainerExit(null, status))
+            .catch((err) => this._handleContainerExit(err));
+    }
+
+    _handleOutput(data) {
+        this.output += data;
+        if (this.output.indexOf('\n') >= 0) {
+            this._callIsFinished();
+        }
+    }
+
+    _handleTimeout() {
+        debug(`enter _timeout(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+        this._checkState([IN_CALL]);
+        this.timeoutID = null;
+        this._cleanup();
+        this.state = EXITING;
+        this._callCallback(new Error('timeout exceeded, killing DockerCaller container'));
```

Comment:
I’ve gotten a bit lost with timeouts 😄 — it looks like the code caller inside the container will also enforce a timeout, which will be cheaper because it won’t require starting a new container. But I’m confused about the timeout durations — is the timeout inside the container shorter than the timeout outside the container?

Reply:
In the latest code, I have the container timeout set at 1.5x the timeout of the "inside the container" timeout - that should help ensure that a timeout almost exclusively means restarting the zygote and not the container itself.